feat(observability): Añadir métricas de Prometheus para monitoreo del…

sebapinery opened the pull request:
… sistema

Métricas añadidas:
- queuety_messages_published_total{topic} - Total published messages
- queuety_messages_delivered_total{topic} - Total delivered messages
- queuety_messages_failed_total{topic} - Failed message deliveries
- queuety_message_processing_seconds{topic, operation} - Processing latency histogram
- queuety_topics_total - Number of active topics
- queuety_subscribers_total{topic} - Subscribers per topic
- queuety_active_connections - Current TCP connections count
- queuety_badger_operations_total{operation, status} - BadgerDB operation metrics
- queuety_auth_attempts_total{result} - Authentication attempts tracking

sebapinery committed:
feat(observability): Añadir métricas de Prometheus para monitoreo del sistema

Métricas añadidas:
- queuety_messages_published_total{topic} - Total published messages
- queuety_messages_delivered_total{topic} - Total delivered messages
- queuety_messages_failed_total{topic} - Failed message deliveries
- queuety_message_processing_seconds{topic, operation} - Processing latency histogram
- queuety_topics_total - Number of active topics
- queuety_subscribers_total{topic} - Subscribers per topic
- queuety_active_connections - Current TCP connections count
- queuety_badger_operations_total{operation, status} - BadgerDB operation metrics
- queuety_auth_attempts_total{result} - Authentication attempts tracking

sebapinery committed:
fix: rollback changes in stats

sebapinery committed:
feat(observability): Añadir soporte para variable de entorno QUEUETY_METRICS_ENABLED

- Implementar lógica para habilitar/deshabilitar métricas de Prometheus
- Por defecto, las métricas están habilitadas
- Se puede deshabilitar configurando QUEUETY_METRICS_ENABLED=false

sebapinery committed:
feat(observability): Add OpenTelemetry and Prometheus metrics instrumentation

- Implement OpenTelemetry tracing for key server operations
- Add spans for database, message handling, and authentication
- Create comprehensive observability section in README
- Configure environment variables for metrics and tracing
- Instrument BadgerDB operations with spans
- Add metrics for messages, performance, and system status

Closes #12

sebapinery committed:
fix: standar env

tomiok reviewed (changes requested):
Si bien hay muchos comentarios, son casi todos de styling, ninguno de logica dura.

Si a cambiar, los comments dentro de las funciones, borrarlos TODOS. Es mas, diria de borrar todos y despues ver  si hace falta agregar alguno.

Al margen de los comments, estoy con muchas ganas de seguir de cerca este PR para que quede optimo y 'mergearlo' a nuestro trunk.

Sigo atento cualquier cambio.

sebapinery committed:
fix: remove comments

sebapinery committed:
fix: translate Observability section in README.md & comments in .env.example

sebapinery committed:
Update server/persistence.go

Co-authored-by: Tomas Francisco Lingotti <tomaslingotti@gmail.com>

sebapinery committed:
feat(observability): Improve span and error handling

- Remove redundant EndSpan calls in multiple methods
- Add panic recovery for methods with spans
- Ensure consistent error and span handling
- Add missing newline characters to log messages

sebapinery committed:
add go.mod & go.sum

tomiok commented on server/server.go:
```diff
@@ -275,42 +375,159 @@ func (s *Server) save(message Message) {
+		"message_ack",
+		observability.WithSpanKind(trace.SpanKindInternal),
+	)
+	defer func() {
+		if r := recover(); r != nil {
+			log.Printf("recovered from panic in ack: %v\n", r)
+			observability.EndSpan(span, fmt.Errorf("panic: %v", r))
+		}
+	}()
+	defer observability.EndSpan(span, nil)
```
entiendo, pero los panics deberian ser casos excepcionales. En todo caso podriamos usar un solo recover a mas alto nivel donde se manejan las conexiones. Tenerlo replicado por todas las funciones es un overkill (creo) pero lo que si se es que no es una practica comun, al menos no lo vi nunca.

sebapinery committed:
feat: refactor and improve distributed tracing system

- Refactor span handling in server.go to eliminate repetitive code
- Add new helper functions in tracing.go to simplify span management
- Implement better context propagation between functions
- Add tracing to save() function for messages
- Improve error handling with RecordSpanError
- Introduce SpanHelper struct to facilitate span management
- Add functions for automatic panic and error handling
- Add Docker Compose configuration for easy deployment
- Add Prometheus configuration for metrics collection

sebapinery committed:
fix: clean docker compose
